File: tang/src/main/cs/Tang/Implementations/SetInjectionPlan.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Com.Microsoft.Tang.Types;
using Com.Microsoft.Tang.Util;

namespace Com.Microsoft.Tang.Implementations
{
    public class SetInjectionPlan : InjectionPlan
    {
        private readonly ISet<InjectionPlan> entries = new MonotonicSet<InjectionPlan>();
        private readonly int numAlternatives;
        private readonly bool isAmbiguous;
        private readonly bool isInjectable;

       public SetInjectionPlan(INode name, ISet<InjectionPlan> entries) : base(name)
       {
            foreach (InjectionPlan ip in entries)
            {
                this.entries.Add(ip);
            }
            int numAlternatives = 1;
            bool isAmbiguous = false;
            bool isInjectable = true;
            foreach (InjectionPlan ip in entries) 
            {
                numAlternatives *= ip.GetNumAlternatives();
                isAmbiguous |= ip.IsAmbiguous();
                isInjectable &= ip.IsInjectable();
            }
            this.numAlternatives = numAlternatives;
            this.isAmbiguous = isAmbiguous;
            this.isInjectable = isInjectable;
        }

       public override int GetNumAlternatives()
       {
           return numAlternatives;
       }

       public override bool IsAmbiguous()
       {
           return isAmbiguous;
       }

       public override bool IsInjectable()
       {
           return isInjectable;
       }

       public override bool HasFutureDependency()
       {
           return false;
       }

       public ISet<InjectionPlan> GetEntryPlans()
       {
           return new MonotonicSet<InjectionPlan>(this.entries);
       }

       public override string ToAmbiguousInjectString() 
        {
            StringBuilder sb = new StringBuilder(GetNode().GetFullName() + "(set) includes ambiguous plans [");
            foreach (InjectionPlan ip in entries) 
            {
                if(ip.IsAmbiguous()) 
                {
                    sb.Append("\n" + ip.ToAmbiguousInjectString());
                }
            }
            sb.Append("]");
            return sb.ToString();
        }

        public override string ToInfeasibleInjectString() 
        {
            StringBuilder sb = new StringBuilder(GetNode().GetFullName() + "(set) includes infeasible plans [");
            foreach (InjectionPlan ip in entries) 
            {
                if(!ip.IsFeasible()) 
                {
                    sb.Append("\n" + ip.ToInfeasibleInjectString());
                }
            }
            sb.Append("\n]");
            return sb.ToString();
        }

        public override bool IsInfeasibleLeaf()
        {
            return false;
        }

        public override string ToShallowString() 
        {
            StringBuilder sb = new StringBuilder("set { ");
            foreach (InjectionPlan ip in entries) 
            {
                sb.Append("\n" + ip.ToShallowString());
            }
            sb.Append("\n } ");
            return null;
        }
    }
}

File: tang/src/main/java/com/microsoft/tang/implementation/SetInjectionPlan.java
package com.microsoft.tang.implementation;

import java.util.Set;

import com.microsoft.tang.types.Node;
import com.microsoft.tang.util.MonotonicHashSet;

public class SetInjectionPlan<T> extends InjectionPlan<T> {
  private final Set<InjectionPlan<T>> entries = new MonotonicHashSet<>();
  private final int numAlternatives;
  private final boolean isAmbiguous;
  private final boolean isInjectable;
  public SetInjectionPlan(Node name, Set<InjectionPlan<T>> entries) {
    super(name);
    this.entries.addAll(entries);
    int numAlternatives = 1;
    boolean isAmbiguous = false;
    boolean isInjectable = true;
    for(InjectionPlan<T> ip : entries) {
      numAlternatives *= ip.getNumAlternatives();
      isAmbiguous |= ip.isAmbiguous();
      isInjectable &= ip.isInjectable();
    }
    this.numAlternatives = numAlternatives;
    this.isAmbiguous = isAmbiguous;
    this.isInjectable = isInjectable;
  }

  @Override
  public int getNumAlternatives() {
    return numAlternatives;
  }

  @Override
  public boolean isAmbiguous() {
    return isAmbiguous;
  }

  @Override
  public boolean isInjectable() {
        return isInjectable;
    }

  @Override
  @Deprecated
  public boolean hasFutureDependency() {
    return false;
  }
  
  public Set<InjectionPlan<T>> getEntryPlans() {
    return new MonotonicHashSet<>(this.entries);
  }
  

  @Override
  protected String toAmbiguousInjectString() {
    StringBuilder sb = new StringBuilder(getNode().getFullName() + "(set) includes ambiguous plans [");
    for(InjectionPlan<T> ip : entries) {
      if(ip.isAmbiguous()) {
        sb.append("\n" + ip.toAmbiguousInjectString());
      }
    }
    sb.append("]");
    return sb.toString();
  }

  @Override
  protected String toInfeasibleInjectString() {
    StringBuilder sb = new StringBuilder(getNode().getFullName() + "(set) includes infeasible plans [");
    for(InjectionPlan<T> ip : entries) {
      if(!ip.isFeasible()) {
        sb.append("\n" + ip.toInfeasibleInjectString());
      }
    }
    sb.append("\n]");
    return sb.toString();
  }

  @Override
  protected boolean isInfeasibleLeaf() {
    return false;
  }

  @Override
  public String toShallowString() {
        StringBuilder sb = new StringBuilder("set { ");
        for(InjectionPlan<T> ip : entries) {
            sb.append("\n" + ip.toShallowString());
        }
        sb.append("\n } ");
        return null;
    }

}
